### DISCUSSION TITLE
Creating Interactive Pie and Donut charts with Plotly

### DISCUSSION THREAD

**Discussion by n-srinidhi** [h2oai/wave#2021] [Q&A] [answered]
Hi!I am trying to create an interactive Pie chart using Plotly, since the Wide Pie Card did not meet my requirements. However, The Frame Card isn't generating, What are the different ways I can display Pie and Donut charts on Wave and keep them updated and interative?(By Interactive I mean Pop ups on click etc.)

```
df=q.client.frame_dict['least_traded']['overall']
fig = px.pie(df, values='Login', names='Symbol', title='Least Traded Symbols')
config = {
        'scrollZoom': False,
        'showLink': False,
        'displayModeBar': False
}
html = pio.to_html(fig, validate=False, include_plotlyjs='cdn', config=config)
# Render the plot as an HTML.
q.page['graph'] = ui.frame_card(box=ui.box(zone='middle',order=2), title='', content=html)
```

Layout is as follows:

```
meta=ui.meta_card(
    box='',
    title='Centra',
    stylesheet=ui.inline_stylesheet(style),
    #scripts=[ui.script('https://cdnjs.cloudflare.com/ajax/libs/Chart.js/4.3.0/chart.min.js')],
    #script=ui.inline_script(content=donut_script,targets=['myChart_do'],),
    layouts=[
        ui.layout(
            breakpoint='xs',
            zones=[
                ui.zone(name='header'),
                ui.zone(name='top',size='20%'),
                ui.zone(name='middle',size='20%'),
                ui.zone(name='bottom',direction=ui.ZoneDirection.ROW,size='60%',zones=[
                    ui.zone('bot_left',size='50%'),
                    ui.zone('bot_right',size='50%')
                ]),
                #ui.zone(name='footer')
                ]),
            ui.layout(
            breakpoint='xl',
            zones=[
                ui.zone(name='header'),
                ui.zone(name='top'),
                ui.zone(name='middle'),
                ui.zone(name='bottom',direction=ui.ZoneDirection.ROW,zones=[
                    ui.zone('bot_left',size='50%'),
                    ui.zone('bot_right',size='50%')
                ]),
                #ui.zone(name='footer')
                ])
            ])
```

`df` is a dataframe with this format:
![image](https://github.com/h2oai/wave/assets/117831190/b3da3f2f-1879-469f-9e8c-7f8751aa7b17)

**Comment by mturoci** [comment]
Are you able to render the plot outside of Wave? E.g. in a plain HTML or Jupyter notebook? Another guess is that there is not enough space in your `middle` zone so adding a fixed `size` could help. However, all these are just guesses unless you provide a runnable minimal repro code.

**Reply by n-srinidhi** [reply]
Can you tell me what additional minimal repro code is required? Meanwhile I'll try executing it in a jupyter notebook.

**Reply by mturoci** [reply]
> Can you tell me what additional minimal repro code is required?

A single piece of code I can copy paste, hit `wave run` and see the problem. What you provided so far is just pieces of code that are not runnable. Also please make sure the minimal code is really minimal and only includes really the necessary parts. E.g. this is not minimal:

```py
meta=ui.meta_card(
    box='',
    title='Centra',
    stylesheet=ui.inline_stylesheet(style),
    #scripts=[ui.script('https://cdnjs.cloudflare.com/ajax/libs/Chart.js/4.3.0/chart.min.js')],
    #script=ui.inline_script(content=donut_script,targets=['myChart_do'],),
    layouts=[
        ui.layout(
            breakpoint='xs',
            zones=[
                ui.zone(name='header'),
                ui.zone(name='top',size='20%'),
                ui.zone(name='middle',size='20%'),
                ui.zone(name='bottom',direction=ui.ZoneDirection.ROW,size='60%',zones=[
                    ui.zone('bot_left',size='50%'),
                    ui.zone('bot_right',size='50%')
                ]),
                #ui.zone(name='footer')
                ]),
            ui.layout(
            breakpoint='xl',
            zones=[
                ui.zone(name='header'),
                ui.zone(name='top'),
                ui.zone(name='middle'),
                ui.zone(name='bottom',direction=ui.ZoneDirection.ROW,zones=[
                    ui.zone('bot_left',size='50%'),
                    ui.zone('bot_right',size='50%')
                ]),
                #ui.zone(name='footer')
                ])
            ])
```

but this is

```py
meta=ui.meta_card(
    box='',
    stylesheet=ui.inline_stylesheet(style)
    layouts=[
        ui.layout(
            ui.layout(
            breakpoint='xl',
            zones=[
                ui.zone(name='middle'),
                ])
            ])
```

**Reply by n-srinidhi** [reply]
@mturoci Hi! So far I got the chart rendered on my dash board using the `ui.frame()` component within a form card. It looks somewhat like this:
![image](https://github.com/h2oai/wave/assets/117831190/f3271ba2-1309-432d-beb3-a1c42ac12899)

My form card code is as follows:
```
most_graph=ui.form_card(
            box=ui.box(zone='top_g_graph'),
            title='',
            items=[
                ui.inline(
                items=[
                ui.text_m(content='Most Popular Symbols'), 
                ui.menu(items=[
                 ui.command(name='tog_pop_o',label='Overall',value='tog_pop_o',),
                 ui.command(name='tog_pop_3',label='Last 3 Months',value='tog_pop_3'),
                 ui.command(name='tog_pop_today',label='Today',value='tog_pop_today')]
                )
                ],align='start',justify='between'
                ),
                ui.frame(content=fig_content,height='300px')
            ],
        )


```
where `fig_content` is the plotly chart code

1. I need to eliminate the scroll bar, basically just fill the chart in the entire region. 
2. I also want to make it transparent i.e. it must fit into the dark theme background, instead of the white default one. 
How can I get this to work? Is there a direct way to do this in Wave or do I need to modify the CSS?

**Reply by mturoci** [reply]
Good job @n-srinidhi!

TLDR: You need to do it via CSS.

More details: Your pie chart is an arbitrary HTML + CSS produced by plotly. Wave is capable of rendering such HTMLs, but (as you noticed) at the cost of harder tweaking and breaking the look&feel consistency. Since this HTML is external, Wave cannot do anything about it so your only option is to tweak it externally, via custom CSS.

Note on pie charts: https://scc.ms.unimelb.edu.au/resources/data-visualisation-and-exploration/no_pie-charts#:~:text=Pies%20and%20doughnuts%20fail%20because,important)%20are%20tricky%20to%20show.

**Reply by n-srinidhi** [reply]
Thank you for your response. Isn't there a better or a simpler way to render donuts and pie charts? The Wide Pie Stat Card doesn't meet my requirements.

**Reply by mturoci** [reply]
There is currently not. In fact, we want to discourage the usage of pie / donut charts as these usually have worse UX than regular bar plots for example.

> The Wide Pie Stat Card doesn't meet my requirements.

What are your requirements? Being able to handle the click event?

**Reply by n-srinidhi** [reply]
Click event yes, Plus it looks kind of flat on the dashboard.

**Reply by mturoci** [reply, marked as answer]
Both issues should be already addressed in our bar charts. However, if you still want those features on the pie chart, you can make a feature request and accompanying PR to show how you envision it.

**Reply by n-srinidhi** [reply]
Alright. Thanks a lot!
